### ISSUE TITLE
Reachability Example not working

### ISSUE THREAD

**Issue by wadoon** [johnyf/dd#31]
Thank you for the great package. 

I encountered following exception: 

```python
bash-4.4$ pwd
/home/weigl/dd-0.5.1/examples
bash-4.4$ python2 reachability.py 
Traceback (most recent call last):
  File "reachability.py", line 56, in <module>
    q = least_fixpoint(transitions, bdd)
  File "reachability.py", line 44, in least_fixpoint
    next_q = _bdd.rename(q, bdd, prime)
  File "/home/weigl/.local/lib/python2.7/site-packages/dd/autoref.py", line 237, in rename
    r = _bdd.rename(u.node, u.bdd, dvars)
  File "/home/weigl/.local/lib/python2.7/site-packages/dd/bdd.py", line 1536, in rename
    assert abs(u) in bdd, u
  File "/home/weigl/.local/lib/python2.7/site-packages/dd/autoref.py", line 45, in __contains__
    assert self is u.bdd
AttributeError: 'int' object has no attribute 'bdd'
```

--- 

dd verion: `0.5.1`, installed via pip
Python: `2.7.13` on a Fedora 26

**Comment by johnyf** [comment]
Thanks for reporting this error, which reminded updating the examples. I pushed to branch `dev` a changeset (2748bb10f0b2b4ab4953dcff21363540c86dd89e) where the examples work.


### Cause of observed error

The reported error arose due to API changes. There are two layers of BDD managers: high and low. For `dd.cudd`, the low level is CUDD itself, the high level is in Cython. For `dd.autoref` both levels are in Python. BDD nodes in low layers are integers in Python (`dd.bdd`) and pointers inside CUDD. The high level abstracts away those implementation details, and there the nodes are `Function` instances, of the corresponding module (and the APIs look almost identical).

`Function`s used to reference only the low level manager `dd.bdd.BDD` that corresponded to them, via the attribute `Function.bdd`. 9935c8eb94e58a406586fe25130dd05ad500d3b7 changes this, renaming that attribute to `Function.manager`. The motivation for this rename was to store in `Function.bdd` the high level manager (in this case an instance of `dd.autoref.BDD`), as done in 5227ab778a35cd314a22f2eea545b89820357785, which addressed (#20). This API change caused the above error.


### How the error has been addressed

This error does not arise on branch `dev` due to a further API change (79f060636069d0c45a0cd4c653fc265a9d27765a) that removes the function `dd.autoref.rename`, in favor of the method `dd.BDD.let` (#12). The example `reachability.py` has been updated on branch `dev` to call the method `let` (which is inspired by the syntax `LET x == 1  y == 2 IN x + y` of [TLA+](https://en.wikipedia.org/wiki/TLA%2B)).


### Interface at the level of integers

The reachability example in `dd` is at the bit level. An interface where variables can be integer-valued or Boolean valued is available in the module [`omega.symbolic.fol`](https://github.com/johnyf/omega/blob/master/omega/symbolic/fol.py). Docstrings document this module for now, but I have been working on updating the package's documentation to provide more information about using this interface.

One step of a reachability computation using `omega.symbolic.fol` looks as follows:

```python
from omega.symbolic import fol
from omega.symbolic import bdd as sym_bdd

c = fol.Context()
d = {"a": (1, 4), "a'": (1, 4), "b": 'bool', "b'": 'bool'}
c.declare(**d)
>>> c.vars
{'a': {'bitnames': ['a_0', 'a_1', 'a_2'],
  'dom': (1, 4),
  'signed': False,
  'type': 'int',
  'width': 3},
 "a'": {'bitnames': ["a_0'", "a_1'", "a_2'"],
  'dom': (1, 4),
  'signed': False,
  'type': 'int',
  'width': 3},
 'b': {'type': 'bool'},
 "b'": {'type': 'bool'}}

# `u` is a BDD node represented by a `Function` instance (of `dd.cudd` if installed, otherwise of `dd.autoref`)
u = c.add_expr("(a = 1) /\ (a' = 2)")  # action that describes allowed transitions
start = c.add_expr('a = 1')  # states from where the transition occurs
r = u & start  # a step from a state that satisfies `start` and the action `u` (an action in TLA+ is a formula that mentions both primed and unprimed variables)
r = c.exist(['a'], r)  # could also write `r = c.add_expr('\E a:  {bdd}'.format(bdd=r))`

>>> r.support  # `Function.support` gives the BDD node's support in bits
{"a_0'", "a_1'", "a_2'"}

>>> c.support(r)  # `Context.support` gives the BDD node's support in high-level vars (integers and Booleans)
{"a'"}

t = sym_bdd.unprime(r, c)  # unprime the variables---can be done also by calling `c.let`

>>> t.support
{'a_0', 'a_1', 'a_2'}

>>> c.support(t)
{'a'}

>>> c.to_expr(t)
'(a = 2)'
```

It may be better to install `omega` from branch `dev`, if using branch `dev` of `dd` (until the new versions are released).


### Travis CI

The Travis CI tests on `dev` fail due to a release candidate of `networkx == 2.0`, which changes its API and will be addressed by #30.

**Comment by johnyf** [comment]
Addressed by c285aae726e3d6f254de51bf1675e8d6c9b05b1b. Please see the CHANGELOG entry for `dd == 0.5.2` for a summary of API changes.
